Basic Documentation and workflow for publishing (#84)
nucypher/nucypher-ts: website/docs/Conditions.md

---
slug: Conditions.md
sidebar_position: 20
---

# Conditions
Several types of access conditions can be defined:
- EVM - on chain state, eg NFT ownership, ETH balance, tx status, contract function call

> - RPC - ethereum RPC calls as defined in the [Official API](https://ethereum.org/en/developers/docs/apis/json-rpc/#json-rpc-methods)

Reviewer: There is a current allow list of methods that limits the calls that can be made, no -https://github.com/nucypher/nucypher/blob/tdec/nucypher/policy/conditions/evm.py#L96?

cc @KPrasch

Reviewer 2: My 2 cents:
- At this moment the EVM condition supports only ERC20 contract (with ERC721 arriving shortly); no testing with custom ABIs has been done
- `nucypher-ts` supports `eth_balance` and `balanceOf` methods (tested).

- Timelock - time based conditions, eg Block Time, Block Height, UTC Time

## EVM Conditions
Here is an example EVM condition for ownership of a specific ERC721 token (NFT):

```js
const ERC721Conditions = {
    contractAddress: "0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D",
    standardContractType: "ERC721",
    chain:  "ethereum",
    method: "ownerOf",
    parameters: [
      "5954"
    ],
    returnValueTest: {

>       comparator: "=",

Reviewer: ```suggestion
      comparator: "==",
```

I suspect this was taken from here - https://github.com/nucypher/nucypher/blob/tdec/tests/data/test_conditions.json#L51 - @KPrasch this test condition should be updated - doesn't seem to fail testing though, maybe unused for now.

- `standardContractType` can take values from `ERC20, ERC721` and `ERC1155`. Alternatively, an ABI can be passed through if a non standard contract is being used.
- `chain` currently only `ethereum` is supported, please [Contact Us](https://discord.gg/RwjHbgA7uQ) if you require non ethereum based conditions.
- `method` the contract method that will be called.
- `parameters` the parameters that will be passed to the contract's `method`.
- `returnValueTest` defines how the return value of the contract call should be evaluated.

In the above example, we query the `ownerOf` method of an `ERC721` token contract which is located at `0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D`.
We are testing whether the `ownerOf` token number `5954` is the current user.
The symbol `:userAddress` is how we define the current user, who will have to authenticate themselves by signing a message using a tool such as MetaMask.

### Ownership of any token in an ERC721 collection (NFT Collection)

```js
const ERC721Conditions = {
    contractAddress: "0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D",
    standardContractType: "ERC721",
    chain:  "ethereum",

>     method: "blanceOf",

Reviewer: ```suggestion
    method: "balanceOf",
```

    parameters: [
      ":userAddress"
    ],
    returnValueTest: {
      comparator: ">",
      value: "0"
    }
  }
```

### Own at least one ERC20 Token
Here we're checking whether the user owns any `T` Threshold Network token
```js
const ERC20Conditions = {
    contractAddress: "0xCdF7028ceAB81fA0C6971208e83fa7872994beE5",
    standardContractType: "ERC20",
    chain:  "ethereum",

>     method: "blanceOf",

Reviewer: ```suggestion
    method: "balanceOf",
```

    parameters: [
      ":userAddress"
    ],
    returnValueTest: {
      comparator: ">",
      value: "0"
    }
  }
```

### Ownership of at least one ERC1155 token from a batch of ids
Batching can be applied to ERC721 tokens as well.
```js

> const ERC115Conditions = {

Reviewer: ```suggestion
const ERC1155Conditions = {
```

    contractAddress: "0x54F456B544abFb785694400bcb1D85629B2D437f",

>     standardContractType: "ERC115",

Reviewer: ```suggestion
    standardContractType: "ERC1155",
```

The function we wish to call is `stakes` which takes an `address` as it's parameter.
We need to provide the `contractAddress`, `functionName`, `functionParams`, and `functionAbi` when defining the condition.

```js
const customABICondition = {
    contractAddress: "00x01B67b1194C75264d06F808A921228a95C765dd7",
    functionName: "stakes",
    functionParams: [":userAddress"],
    functionAbi: {
        inputs: [
        {
            internalType: "address",
            name: "stakingProvider",
            type: "address"
        }
        ],
        name: "stakes",
        outputs: [
        {
            internalType: "uint96",
            name: "tStake",
            type: "uint96"
        },
        {
            internalType: "uint96",
            name: "keepInTStake",
            type: "uint96"
        },
        {
            internalType: "uint96",
            name: "nuInTStake",
            type: "uint96"
        }
        ],
        stateMutability: "view",
        type: "function"
    },
    chain:  "ethereum",
    returnValueTest: {
      key: "tStake",
      comparator: ">",
      value: "0"
    }
  }
```

## RPC Conditions

> Here will will query the ETH balance of the users address using the RPC call [`eth_getBalance`](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getbalance)

Reviewer: ```suggestion
Here we will query the ETH balance of the user's address using the RPC call [`eth_getBalance`](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getbalance)
```

```js
const ETHBalance = {
    contractAddress: "",
    standardContractType: "",
    chain:  "ethereum",
    method: "eth_getBalance",
    parameters: [
      ":userAddress",
      "latest"
    ],
    returnValueTest: {
      comparator: ">=",
      value: "10000000000000"
    }
  }
```

> ## Time Conditions

Reviewer: I guess we are waiting to write this up until the code is changed to use blockTime instead of system time?

Author: exactly
